core/src/main/java/com/linecorp/armeria/client/retry/RetryRequestStrategyByStatus.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpObject;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.common.http.HttpStatusClass;
+
+/**
+ * Provides the {@link RetryRequestStrategy} that decides to retry the request again using HTTP statuses.
+ */
+public class RetryRequestStrategyByStatus implements RetryRequestStrategy<HttpRequest, HttpResponse> {
```

Comment:
Factory method part is a good idea. With package-private part, I doubt about it. Because a client may want extend this class and override shouldRetry(request, throwable) method.